proposals/ideas/xxx-flashflow.txt

```diff
@@ -0,0 +1,950 @@
+
+Thus the odds the adversary can cheat in `X` consecutive buckets are
+
+    [(bucket_size-1)/bucket_size]^X
+
+In our case, X will be highly varied: Slow relays won't see very many
+buckets, but fast relays will. The damage to the network a very slow relay can
+do by faking being only slightly faster is limit. Nonetheless, for now we
+motivate the selection of bucket size, N, with a slow relay:
+
+  - Assume a very slow relay of 1 Mbit/s capacity that will cheat 1 cell in
+    each bucket. Assume a 30 second measurement.
+  - The relay will handle `1 * 30` = 30 Mbit of traffic during the
+    measurement, or 3.75 MB, or 3.75 million bytes.
+  - Cells are 514 bytes. Approximately (e.g. ignoring TLS) 7300 cells will be
+    sent/recv over the course of the measurement.
+  - A `bucket_size` of 50 results in about 146 buckets over the course of the
+    30s measurement.
+  - Therefore, the odds of the adversary cheating successfully as
+    `(49/50)^(146)`, or about 5.2%.
+
+This sounds high, but a relay capable of double the bandwidth (2 Mbit/s) will
+have (49/50)^(2*146) or 0.2% odds of success. Quite low already.
+
+Wanting a <1% chance that a 10 Mbit/s relay can successfully cheat results in
+a bucket size of approximately 125:
+
+  - 10 * 30 = 300 Mbit of traffic during 30s measurement. 37.5 million bytes.
+  - 37,500,000 bytes / 514 bytes/cell = ~73,000 cells
+  - bucket size of 125 cells means 73,000 / 125 = 584 buckets
+  - (124/125)^(584) = 0.918% chance of successfully cheating
+
+Slower relays can cheat more easily, faster relays are less likely to be able
+to cheat.
+
+X. FlashFlow measurement system: Medium term
+
+FlashFlow has been implemented and relays are starting to update to a version
+of Tor that supports it.
+
+NRL will host a FlashFlow deployment consisting of a FF coordinator and a
+single FF measurer on a single 1 Gbit/s machine. Data produced by this
+deployment will be made available (semi?) publicly, including both v3bw files
+and intermediate results.
+
+X. FlashFlow measurement system: Long term
+
+In the long term, finishing-touch development work will be done and
+experiments will be run to determine the best way to integrate FlashFlow into
+the Tor ecosystem.
+
+X.X Authentication to Target Relay
+
+FlashFlow measurers will maintain TLS identity certificates and use them when
+connecting to target Tor relays. Dirauths will vote on which measurers are
+allowed to perform measurements. Relays will update from stub authentication
+methods to methods that actually check that the measurer used a client TLS
+certificate and that it's listed in the consensus.
+
+X.X Experiments
+
+   [XXX todo]
+
+X.X Other Changes/Investigations/Ideas
+
+- Add a cell that the target relay sends to the measurers indicating its CPU
+  and memory usage, whether it has a shortage of sockets, how much bandwidth
+  load it has been experiencing lately, etc. Use this information to lower a
+  relays weight, never increase.
```

Comment:
Please explicitly and directly add in that list that Mike pasted to the email thread I started like 5 different times.

Reply:
really good idea